[TensorExpr] Add LLVM codegen.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34234 [WIP] Changes in test/test_jit_fuser.py test/test_jit_fuser_te.py.
* #34233 [WIP] Changes to peephole.cpp.
* #34231 [WIP] Guard elimination changes from bertmaher/pytorch_fusion.
* #34230 [TensorExpr] Add tensorexpr benchmarks.
* **#34228 [TensorExpr] Add LLVM codegen.**
* #34227 [TensorExpr] Add CUDA codegen.
* #34226 [TensorExpr] Add a fuser pass based on tensor expressions.
* #34224 [TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch.

This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

Commit:
Add LLVM codegen for tensor expressions.

[ghstack-poisoned]

Commit:
Update on "Add LLVM codegen for tensor expressions."


Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update descriptions. on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Lints on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

ezyang commented on caffe2/CMakeLists.txt:
```diff
@@ -487,6 +489,21 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
 
+  if (USE_LLVM)
+    message(STATUS "Looking for LLVM in ${USE_LLVM}")
+    find_package(LLVM QUIET PATHS ${USE_LLVM} NO_DEFAULT_PATH)
```
Most of the `find_package` live in `cmake/Dependencies.cmake`. I don't *really* care where you put it, but it might make it harder for people to find this logic in the future.

ezyang commented on caffe2/CMakeLists.txt:
```diff
@@ -487,6 +489,21 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
+      message(STATUS "Using LLVMConfig.cmake in: ${LLVM_DIR}")
+
+      include_directories(${LLVM_INCLUDE_DIRS})
+      add_definitions(-DENABLE_LLVM ${LLVM_DEFINITIONS})
```
Uhhh, as far as I can tell, `LLVM_DEFINITIONS` isn't used anywhere. Did you add a usage in a different PR?

Note that adding `-D` flags is an antipattern unless you can guarantee that `ENABLE_LLVM` only is in c++ files, as in general clients compiling against your library won't set this definition. But I see usage in headers... you should put this in one of our Config.h headers in that case

ezyang replied:
And if you do put it in a Config.h, please prefix it with TORCH following our other definitions

bertmaher replied:
Would it be better to just conditionally include the llvm-related sources?  The only thing we really do with ENABLE_LLVM is to #ifdef out entire files.

ezyang replied:
I'm not sure what you mean by "conditionally include"

bertmaher replied:
Oh, I mean in cmakelists, do something like 
if (USE_LLVM)
list(append TORCH_SRCS llvm_codegen.cpp ...)
endif()

ZolotukhinM replied:
One important thing to note is that nothing internally hard depends on LLVM and LLVM is not exposed in any way to the user of libtorch. However, if libtorch was build with USE_LLVM, it gets an LLVM codegen for tensor expression engine (the codegen is registered via static constructors and is picked dynamically based on what is available). In that sense, I don't think it's important for whoever links with pytorch to get the value of `ENABLE_LLVM` used when pytorch was built.

ZolotukhinM replied:
That said, I'll give it a shot to register LLVM as a usual dependency.

ezyang replied:
@bertmaher So, I am perfectly happy to believe that things will "work out ok" the way things are setup today. What I am not convinced about, is when people who are not you start touching this code, they might start including the header from different places... some of which are publicly accessible... and then weird things start breaking and they don't know why. Unless you are going to add a lint rule enforcing whatever informal rule you've specified here ("nothing internally hard depends on LLVM and LLVM is not exposed in any way to the user of libtorch"), it's better to arrange for it to just be impossible to violate this internal rule, or make the consequences of violating it negligible (put it in a generated Config.h header).

bertmaher replied:
@ezyang I'm confused -- I didn't assert that everything will "work out ok" the way things are set up, and I wasn't suggesting we leave it alone.  I noticed that there are several other places in caffe2/CMakeLists.txt where we conditionally append to TORCH_SRCS based on other configuration variables, and I wondered if that was a better approach than using any #ifdefs at all.

ezyang replied:
When I said "work out ok", I was referring specifically to this comment from @ZolotukhinM 

>  In that sense, I don't think it's important for whoever links with pytorch to get the value of ENABLE_LLVM used when pytorch was built.

Sorry about the confusing mention. 

> I noticed that there are several other places in caffe2/CMakeLists.txt where we conditionally append to TORCH_SRCS based on other configuration variables, and I wondered if that was a better approach than using any #ifdefs at all.

This seems better for now, at least.

bertmaher replied:
Ahh thanks for the clarification @ezyang :-).  That makes sense to me.

ZolotukhinM replied:
> Would it be better to just conditionally include the llvm-related sources? The only thing we really do with ENABLE_LLVM is to #ifdef out entire files.

@bertmaher, we actually also use it for selecting a codegen for CPU in kernel.cpp.

ezyang commented on caffe2/CMakeLists.txt:
```diff
@@ -487,6 +489,21 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
+    endif (LLVM_FOUND)
+  endif (USE_LLVM)
+
+  set_source_files_properties(${TORCH_SRC_DIR}/csrc/jit/tensorexpr/llvm_jit.cpp PROPERTIES COMPILE_FLAGS -fno-rtti)
```
Why?

bertmaher replied:
I think this isn't needed anymore.  Early on, some of the llvm classes used by this file ended up missing key methods (since llvm itself is compiled without rtti).  But changes since then have made this unnecessary.  (We've switched to LLJIT and avoided a pile of problematic classes.)

ezyang commented:
NACKing cmake changes, mostly due to unsafe use of `-D` addition to compilation

bertmaher reviewed (approved):
I've reviewed all the llvm stuff in our side branch so I'm comfortable giving this an approval.  Maybe take care of the cmake changes before landing.

Commit:
Rebase and try to fix windows build. on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
LLVM cmake integration changes on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

ZolotukhinM commented:
@ezyang please take another look - I moved the `USE_LLVM` check to `cmake/Dependencies.cmake` and removed the `-fno-rtti` flag.

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add LLVM codegen."


This PR adds LLVM codegen to tensor expressions. LLVM is added as an
optional build dependency specified with `USE_LLVM=<path_to_llvm>`
variable. If this variable is not set or LLVM is not found in the
specified path, the LLVM codegen is completely disabled.

Differential Revision: [D20251832](https://our.internmc.facebook.com/intern/diff/D20251832)

[ghstack-poisoned]
